lib/Dialect/Seq/Transforms/SinkClockGates.cpp

```diff
@@ -0,0 +1,260 @@
+  using SinkClockGatesBase::SinkClockGatesBase;
+
+  // TODO: Can be moved to some utils !!
+  /// Append inputs to the instance, and also update the referenced modules.
+  void appendInputsToInstance(HWInstanceLike instance,
+                              ArrayRef<std::pair<StringAttr, Value>> inputs,
+                              ArrayRef<HWModuleOp> refModules) {
+
+    SmallVector<Attribute> argNames;
+    if (auto args = instance->getAttrOfType<ArrayAttr>("argNames"))
+      argNames.insert(argNames.begin(), args.begin(), args.end());
+
+    SmallVector<Value> newOperands;
+    newOperands.reserve(inputs.size());
+    SmallVector<std::tuple<StringAttr, Type, Location>> newInputs;
+    newInputs.reserve(inputs.size());
+    for (const auto &[name, value] : inputs) {
+      // If no argNames attribute, donot add it.
+      if (!argNames.empty())
+        argNames.push_back(name);
+      // Additional operands to the Instance op.
+      newOperands.push_back(value);
+      // Prepare the additional input ports that need to be added to the
+      // referenced module.
+      newInputs.emplace_back(name, value.getType(), value.getLoc());
+    }
+    // Add the new operands to the instance.
+    instance->insertOperands(instance->getNumOperands(), newOperands);
+    if (!argNames.empty())
+      instance->setAttr("argNames",
+                        ArrayAttr::get(instance->getContext(), argNames));
+    // Update all the modules with the new input ports.
+    for (auto refMod : refModules) {
+      refMod.appendPorts(newInputs, {});
+    }
+  }
+
+  /// Get the referenced modules for an instance.
+  SmallVector<HWModuleOp, 1> getReferencedModules(HWInstanceLike inst) {
+    SmallVector<HWModuleOp, 1> refMods;
+    for (auto refModName : inst.getReferencedModuleNamesAttr()) {
+      auto *node = graph->lookup(cast<StringAttr>(refModName));
+      if (auto instModule = dyn_cast_or_null<HWModuleOp>(*node->getModule()))
+        refMods.push_back(instModule);
+    }
+    return refMods;
+  }
+
+  /// If the `argNames` attribute exists, return it.
+  SmallVector<StringAttr> getInstanceArgNames(HWInstanceLike inst) {
+    SmallVector<StringAttr> argNames;
+    if (auto args = inst->getAttrOfType<ArrayAttr>("argNames")) {
+      auto attr = args.getAsRange<StringAttr>();
+      argNames.insert(argNames.begin(), attr.begin(), attr.end());
+    }
+    return argNames;
+  }
+
+  /// Combine two consecutive clock_gate ops, into one.
+  void collapseConsecutiveClockGates(ClockGateOp driverGate,
+                                     ClockGateOp userGate) {
+    mlir::OpBuilder builder(context);
+    builder.setInsertionPoint(userGate);
+    Value enable = getEnable(driverGate);
+    enable = builder.create<comb::AndOp>(userGate->getLoc(), enable,
+                                         userGate.getEnable());
+    userGate.getInputMutable().assign(driverGate.getInput());
+    userGate.getEnableMutable().assign(enable);
+  }
```

Comment:
This is very neat! Might also make sense as a canonicalization on `ClockGateOp`.

Reply:
Good idea, will add another PR to add this pattern.